File: src/engine/queue.ts
import { db } from '@/data/db'
import { repo } from '@/data/repo'
import type { Card, Deck } from '@/data/types'

export const dayKey = (d: Date = new Date()) =>
  `${d.getFullYear()}-${String(d.getMonth() + 1).padStart(2, '0')}-${String(d.getDate()).padStart(2, '0')}`
const counterKey = (deckId: string) => `newIntroduced:${deckId}:${dayKey()}`

async function deletedNoteIds(): Promise<Set<string>> {
  const notes = await db.notes.filter(n => n.deletedAt != null).toArray()
  return new Set(notes.map(n => n.id))
}

async function liveDecks(deckId?: string): Promise<Deck[]> {
  if (deckId) {
    const deck = await db.decks.get(deckId)
    return deck && deck.deletedAt == null ? [deck] : []
  }
  return db.decks.filter(d => d.deletedAt == null).toArray()
}

async function eligibleCards(deckId?: string): Promise<Card[]> {
  const deadNotes = await deletedNoteIds()
  const cards = await db.cards.toArray()
  const noteById = new Map((await db.notes.toArray()).map(n => [n.id, n]))
  const deadDecks = new Set((await db.decks.filter(d => d.deletedAt != null).toArray()).map(d => d.id))
  return cards.filter(c => {
    if (c.deletedAt != null) return false
    if (c.suspended === 1) return false
    if (deadNotes.has(c.noteId)) return false
    const note = noteById.get(c.noteId)
    if (!note) return false
    if (deadDecks.has(note.deckId)) return false
    if (deckId && note.deckId !== deckId) return false
    return true
  })
}

async function newAllowance(deckId: string): Promise<number> {
  const deck = await db.decks.get(deckId)
  if (!deck) return 0
  const introduced = (await repo.getMeta<number>(counterKey(deckId))) ?? 0
  return Math.max(0, deck.newPerDay - introduced)
}

export async function dueCountsAll(now = Date.now()): Promise<Map<string, { due: number; newAvailable: number }>> {
  const decks = await liveDecks()
  const cards = await eligibleCards()
  const noteById = new Map((await db.notes.toArray()).map(n => [n.id, n]))

  const counts = new Map<string, { due: number; newAvailable: number }>()
  const newInDeck = new Map<string, number>()
  for (const deck of decks) {
    counts.set(deck.id, { due: 0, newAvailable: 0 })
    newInDeck.set(deck.id, 0)
  }

  for (const c of cards) {
    const deckId = noteById.get(c.noteId)?.deckId
    if (deckId == null) continue
    const entry = counts.get(deckId)
    if (!entry) continue
    if (c.state === 0) newInDeck.set(deckId, (newInDeck.get(deckId) ?? 0) + 1)
    else if (c.due <= now) entry.due += 1
  }

  for (const deck of decks) {
    const introduced = (await repo.getMeta<number>(counterKey(deck.id))) ?? 0
    const allowance = Math.max(0, deck.newPerDay - introduced)
    counts.get(deck.id)!.newAvailable = Math.min(allowance, newInDeck.get(deck.id) ?? 0)
  }

  return counts
}

export async function dueCounts(deckId?: string): Promise<{ due: number; newAvailable: number }> {
  const all = await dueCountsAll()
  if (deckId) return all.get(deckId) ?? { due: 0, newAvailable: 0 }
  let due = 0
  let newAvailable = 0
  for (const entry of all.values()) {
    due += entry.due
    newAvailable += entry.newAvailable
  }
  return { due, newAvailable }
}

export async function buildQueue(deckId?: string): Promise<Card[]> {
  const now = Date.now()
  const cards = await eligibleCards(deckId)
  const decks = await liveDecks(deckId)
  const noteById = new Map((await db.notes.toArray()).map(n => [n.id, n]))

  const learning = cards.filter(c => (c.state === 1 || c.state === 3) && c.due <= now).sort((a, b) => a.due - b.due)
  const review = cards.filter(c => c.state === 2 && c.due <= now).sort((a, b) => a.due - b.due)

  const newCards: Card[] = []
  for (const deck of decks) {
    const allowance = await newAllowance(deck.id)
    if (allowance <= 0) continue
    const candidates = cards
      .filter(c => c.state === 0 && noteById.get(c.noteId)?.deckId === deck.id)
      .sort((a, b) => a.due - b.due)
      .slice(0, allowance)
    newCards.push(...candidates)
  }

  return [...learning, ...review, ...newCards]
}

export async function noteNewIntroduced(deckId: string, n: number): Promise<void> {
  const current = (await repo.getMeta<number>(counterKey(deckId))) ?? 0
  await repo.setMeta(counterKey(deckId), current + n)
}

File: ios/Elbert/Engine/Queue.swift
import Foundation
import SwiftData

/// What to study, in what order, and how much of it there is.
///
/// Port of `src/engine/queue.ts`. The soft-delete filtering the web version needs is gone, since
/// deleting a note or deck now cascades to its cards for real, but the orphan checks stay: without
/// a unique constraint or a foreign key the CloudKit mirror can hand over a card whose note never
/// arrived, and a card with no note has nothing to show.
enum Queue {
    /// What a deck has waiting.
    struct Counts: Equatable, Sendable {
        /// Cards past their due time. New cards are never counted here, however overdue their
        /// nominal due date, because they have not been introduced yet.
        var due: Int = 0
        /// New cards that could be introduced right now: the smaller of what the deck has left in
        /// its daily allowance and what it actually holds.
        var newAvailable: Int = 0

        var total: Int { due + newAvailable }
        var isEmpty: Bool { total == 0 }
    }

    // MARK: - Eligibility

    /// Every card that could be studied at all, ignoring due dates.
    ///
    /// Excluded: suspended cards, cards whose note is missing, and cards whose note has no deck.
    static func eligibleCards(
        in context: ModelContext,
        deckID: UUID? = nil
    ) throws -> [Card] {
        try context.fetch(FetchDescriptor<Card>()).filter { card in
            guard !card.suspended, let note = card.note, let deck = note.deck else { return false }
            guard let deckID else { return true }
            return deck.id == deckID
        }
    }

    // MARK: - Counts

    /// Counts for every deck, keyed by deck id. Decks with nothing waiting are still present, with
    /// zeroes, so a caller can render a full deck list from this alone.
    static func counts(
        in context: ModelContext,
        now: Date = Date(),
        counter: NewCardCounter = NewCardCounter()
    ) throws -> [UUID: Counts] {
        let decks = try context.fetch(FetchDescriptor<Deck>())
        let cards = try eligibleCards(in: context)

        var counts: [UUID: Counts] = [:]
        var newInDeck: [UUID: Int] = [:]
        for deck in decks {
            counts[deck.id] = Counts()
            newInDeck[deck.id] = 0
        }

        for card in cards {
            guard let deckID = card.note?.deck?.id, counts[deckID] != nil else { continue }
            if card.state == .new {
                newInDeck[deckID, default: 0] += 1
            } else if card.due <= now {
                counts[deckID]?.due += 1
            }
        }

        for deck in decks {
            counts[deck.id]?.newAvailable = min(
                counter.allowance(for: deck, on: now),
                newInDeck[deck.id] ?? 0
            )
        }

        return counts
    }

    /// Counts for one deck, or for everything at once when `deckID` is nil.
    static func counts(
        in context: ModelContext,
        deckID: UUID?,
        now: Date = Date(),
        counter: NewCardCounter = NewCardCounter()
    ) throws -> Counts {
        let all = try counts(in: context, now: now, counter: counter)

        if let deckID {
            return all[deckID] ?? Counts()
        }
        return all.values.reduce(into: Counts()) { total, entry in
            total.due += entry.due
            total.newAvailable += entry.newAvailable
        }
    }

    // MARK: - The queue

    /// The study queue, in order.
    ///
    /// Learning and relearning cards that are due come first, because they are mid-acquisition and
    /// waiting costs more there than anywhere else. Then reviews that are due, oldest first. Then
    /// new cards, up to each deck's remaining allowance for today.
    ///
    /// Nothing not yet due appears at all: a queue that hands over a card early is quietly
    /// undoing the schedule.
    static func build(
        in context: ModelContext,
        deckID: UUID? = nil,
        now: Date = Date(),
        counter: NewCardCounter = NewCardCounter()
    ) throws -> [Card] {
        let cards = try eligibleCards(in: context, deckID: deckID)

        let learning = cards
            .filter { $0.state.isLearning && $0.due <= now }
            .sorted(by: dueThenID)

        let review = cards
            .filter { $0.state == .review && $0.due <= now }
            .sorted(by: dueThenID)

        var new: [Card] = []
        for deck in try decks(in: context, deckID: deckID) {
            let allowance = counter.allowance(for: deck, on: now)
            guard allowance > 0 else { continue }

            new += cards
                .filter { $0.state == .new && $0.note?.deck?.id == deck.id }
                .sorted(by: dueThenID)
                .prefix(allowance)
        }

        return learning + review + new
    }

    // MARK: - Helpers

    private static func decks(in context: ModelContext, deckID: UUID?) throws -> [Deck] {
        let all = try context.fetch(FetchDescriptor<Deck>())
        guard let deckID else { return all }
        return all.filter { $0.id == deckID }
    }

    /// Due date, then id. The tie-break is not cosmetic: two cards created in the same millisecond
    /// are common, and a queue that shuffles them between builds makes a session feel unstable.
    private static func dueThenID(_ left: Card, _ right: Card) -> Bool {
        if left.due != right.due { return left.due < right.due }
        return left.id.uuidString < right.id.uuidString
    }
}
